q2_annotate/kraken2/merge.py

```diff
@@ -0,0 +1,420 @@
+                else:
+                    if filename in mapping[sample_id]:
+                        msg = (
+                            'Two MAGs with the same uuid were detected '
+                            f'Duplicated uuid: {filename}.'
+                        )
+                        raise ValueError(msg)
+
+                    mapping[sample_id][filename] = format
+        else:
+            filepath = value
+            format = Format(filepath, mode='r')
+            if sample_id not in mapping:
+                mapping[sample_id] = [format]
+            elif minimizers_present:
+                msg = (
+                    'Two or more reports with the same sample ID were '
+                    'attempted to be merged but the option to capture minimzer '
+                    'data was enabled. It is not possible to merge kraken2 '
+                    'reports that contain minimizer information.'
+                )
+                raise ValueError(msg)
+            else:
+                mapping[sample_id].append(format)
+
+    report_mapping = {}
+    for sample_id, value in chained_reports:
+        _update_mapping(
+            sample_id, value, report_mapping, Kraken2ReportFormat
+        )
+
+    output_mapping = {}
+    for sample_id, value in chained_outputs:
+        _update_mapping(
+            sample_id, value, output_mapping, Kraken2OutputFormat
+        )
+
+    return report_mapping, output_mapping
+
+
+def _check_for_minimizers(reports: Kraken2ReportDirectoryFormat) -> bool:
+    '''
+    Checks whether the reports being processed include the optional minimizer
+    columns. This determines whether reports with a shared sample ID can be
+    merged.
+
+    Parameters
+    ----------
+    reports : Kraken2ReportDirectoryFormat
+        The kraken2 reports to examine for minimzer columns.
+
+    Returns
+    -------
+    bool
+        Wether minimizer columns are present in the reports.
+    '''
+    _, report = list(reports.reports.iter_views(Kraken2ReportFormat))[0]
+
+    return 'n_read_minimizers' in report.view(pd.DataFrame)
+
+
+def _merge_reports(
+    reports: list[Kraken2ReportFormat]
+) -> Kraken2ReportFormat:
+    '''
+    Merges two or more kraken2 reports into a single report. See
+    `_merge_trees` for the actual merging algorithm.
+
+    Parameters
+    ----------
+    reports : list[Kraken2ReportFormat]
+        Kraken2 reports belonging to the same sample ID.
+
+    Returns
+    -------
+    Kraken2ReportFormat
+        The merged kraken2 report format.
+    '''
+    if len(reports) == 1:
+        return reports[0]
+
+    merged: tuple[TreeNode | None, TreeNode | None] | None = None
+    while reports:
+        first = _report_df_to_tree(reports.pop().view(pd.DataFrame))
+        if merged is None:
+            second = _report_df_to_tree(reports.pop().view(pd.DataFrame))
+        else:
+            second = merged
+
+        merged = _merge_trees(first, second)
```

Comment:
This should do the same thing with less effort:
```suggestion
    functools.reduce(_merge_trees, 
                     _report_df_to_tree(x.view(pd.DataFrame)) for x in reports)
```